File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/BinaryNumberFormat.java
package com.cedarsoft.commons.javafx;

import java.text.FieldPosition;
import java.text.NumberFormat;
import java.text.ParsePosition;

/**
 * Formats integers and long values to/from binary string
 */
public class BinaryNumberFormat extends NumberFormat {

  private final int minTextLength;

  public BinaryNumberFormat(int minTextLength) {
    this.minTextLength = minTextLength;
  }

  @Override
  public StringBuffer format(double number, StringBuffer toAppendTo, FieldPosition pos) {
    throw new IllegalArgumentException("doubles not supported.");
  }

  @Override
  public StringBuffer format(long number, StringBuffer toAppendTo, FieldPosition pos) {
    StringBuffer stringBuffer = toAppendTo.append("0b");
    String binaryString = Long.toBinaryString(number);
    int padCount = minTextLength - binaryString.length();
    for (int i = 0; i < padCount; i++) {
      stringBuffer.append('0');
    }
    stringBuffer.append(binaryString);
    return stringBuffer;
  }

  @Override
  public Number parse(String source, ParsePosition parsePosition) {
    long l = 0;
    if (source.startsWith("0b")) {
      try {
        l = Long.parseUnsignedLong(source.substring(2), 2);
      }
      catch (NumberFormatException e) {
        e.printStackTrace();
      }
      return l;
    }

    try {
      l = Long.parseUnsignedLong(source, 2);
    }
    catch (NumberFormatException e) {
      e.printStackTrace();
    }
    return l;
  }
}

File: commons/javafx/src/main/kotlin/it/neckar/open/javafx/BinaryNumberFormat.kt
package it.neckar.open.javafx

import java.text.FieldPosition
import java.text.NumberFormat
import java.text.ParsePosition

/**
 * Formats integers and long values to/from binary string
 */
class BinaryNumberFormat(private val minTextLength: Int) : NumberFormat() {
  override fun format(number: Double, toAppendTo: StringBuffer, pos: FieldPosition): StringBuffer {
    throw IllegalArgumentException("doubles not supported.")
  }

  override fun format(number: Long, toAppendTo: StringBuffer, pos: FieldPosition): StringBuffer {
    val stringBuffer = toAppendTo.append("0b")
    val binaryString = java.lang.Long.toBinaryString(number)
    val padCount = minTextLength - binaryString.length
    for (i in 0 until padCount) {
      stringBuffer.append('0')
    }
    stringBuffer.append(binaryString)
    return stringBuffer
  }

  override fun parse(source: String, parsePosition: ParsePosition): Number {
    if (source.startsWith("0b")) {
      try {
        return java.lang.Long.parseUnsignedLong(source.substring(2), 2)
      } catch (e: NumberFormatException) {
        e.printStackTrace()
      }
    }
    try {
      return java.lang.Long.parseUnsignedLong(source, 2)
    } catch (e: NumberFormatException) {
      e.printStackTrace()
    }
    return 0
  }
}
